job controller test: avoid edge condition of clock advance and `DeadlineExceeded` condition detection

harshanarayana opened the pull request:
#### What type of PR is this?
/kind flake

#### What this PR does / why we need it:
This PR Fixes the `TestSyncPastDeadlineJobFinished` UTs for the `job_controller` that were causing the test flakes after the changes from #110710  was merged to the mainline.

#### Which issue(s) this PR fixes:
Fixes #110697

#### Special notes for your reviewer:
Instead of invoking the `manager.Run` to start the controller, switching to use the `manager.syncJob` call directly which give you a more controlled way of processing the reconciliation as well as ability to validate the conditions easily.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
NONE
```

harshanarayana commented:
/test pull-kubernetes-unit

mimowo commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -1840,8 +1840,6 @@ func TestSyncPastDeadlineJobFinished(t *testing.T) {
 	sharedInformerFactory.Start(ctx.Done())
 	sharedInformerFactory.WaitForCacheSync(ctx.Done())
 
-	go manager.Run(ctx, 1)
```
Is the change about replacing `go manager.Run(ctx, 1)` with invocations of `syncJob` necessary? IIUC the issue is that the job has start time shifted by one second to the past and this is exactly the value for `ActiveDeadlineSeconds`, so avoiding being on the edge should be enough. If this is the case then I suggest to make the fix minimal and only keep setting `ActiveDeadlineSeconds` to 2 and doing the sleep `manager.clock.Sleep(2 * time.Second)`.

harshanarayana replied:
>Is the change about replacing go manager.Run(ctx, 1) with invocations of syncJob necessary?

@mimowo not really a mandatory change. But it enables a bit more control on performing the sync is what my thought behind the change was. Do you think that is an overkill ?

mimowo replied:
I wouldn't call it overkill as this is a simple change, but it makes harder to reason about what is needed to fix the test. 

That being said, the change is reasonable to me in a unit test, so I would suggest one of the two approaches:
1. Leave as is, but make it clear in the issue description what is the needed change and what is a drive-by change to make debugging / reasoning about the test simpler. Preferably split the changes into two commits withing the PR.
2. Make only necessary changes in this PR, but open a follow up PR to improve debuggability / clarity of the test.

If you make changes to improve debuggability of the test you may want as well to return the `err` here https://github.com/kubernetes/kubernetes/blob/7eb9a75bf914428e7579d2497722f71be3ef1812/pkg/controller/job/job_controller_test.go#L1900 and log it there: https://github.com/kubernetes/kubernetes/blob/7eb9a75bf914428e7579d2497722f71be3ef1812/pkg/controller/job/job_controller_test.go#L1908

harshanarayana replied:
/Ack.

harshanarayana replied:
@mimowo  I have taken up your suggestion and did what you suggested in option 1. PTAL when possible.

alculquicondor replied:
Please revert this.
If you call syncJob directly, then we are not testing the reinsertions into the workqueue, which is where the original bug was initially.

mimowo commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -1860,8 +1858,9 @@ func TestSyncPastDeadlineJobFinished(t *testing.T) {
 	}
 	for _, tc := range tests {
 		t.Run(tc.name, func(t *testing.T) {
+			jobKey := fmt.Sprintf("default/%s", tc.jobName)
```
Use a helper method `controller.FuncKey` to get the key (but preferably as mentioned about keep the fix minimal if possible).

harshanarayana replied:
/Ack

mimowo reviewed (commented):
IIUC the refactoring to use `syncJob` is not necessary to fix the issue. If this is the case I suggest to keep the fix minimal.

The PR title is not clear to me, I suggest something more to the point, like "Deflake the job controller test: TestSyncPastDeadlineJobFinished".

harshanarayana committed:
job controller test: avoid edge condition due to shift of clock by 1s

In Job controller test, when the `setStartTime` attribute for the test
is configured, we shift the clock back by `1s` and set this value
as the start time for the job while creating the resource. This causes
edge condition where some time the test detects that the `DeadlineExceeded`
condition has been set too early and causes the test flakes.

Once the test is past this phase, during the next iteration, we see that
the fake clock is advanced to cause the deadline condition. However, if
this time just happens to fall short of the mark, then this doesn't set
the `DeadlineExceeded` condition and since we do not advance the clock
after doing so once, the test will fail as the condition for
`DeadlineExceeded` will never get set. This change address the case by
setting the `ActiveDeadlineSeconds` for the job to be `2s` instead of
`1s` which is the internval for which the clock will be moved backwards.

The invocation of `manager.Run` has been replaced with repeated calls to
the `manager.syncJob` call by passign the job Key as the argument in
order to get a more controlled behavior. This also helps avoid
performing an invocation of `wait.PollImmediate` since the call to the
sync is explicit.

Along with the above change there is a `manager.clock.Sleep` introduced
at the tail end of the `wait.PollImmedicate` while checking for the
existence of the `DeadlineExceeded` condition. This was done in order to
avoid certain corner cases that were noticed while performing the test
under load + with `-count 100`.

mimowo commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -1893,15 +1899,23 @@ func TestSyncPastDeadlineJobFinished(t *testing.T) {
+				// Advance the clock again. The fact that in the first pass the flow got up to here but the
+				// condition was not set means that you have hit an edge condition. So advance the clock
+				// here before the controller.syncJob is called again in the next polling interval.
+				manager.clock.Sleep(200 * time.Millisecond)
```
Is there any case this is needed? I'm wondering that since you induce `syncJob` the periodic polling may not be needed at all - the first poll should always detect the new condition. If this is the case I would suggest removing the periodic poll entirely.

harshanarayana replied:
I have seen this happen a few times while testing things locally. Even after setting the active deadline seconds to 2 seconds and sleep to 2 second to advance the clock. 

```bash
 690   │ === RUN   TestSyncPastDeadlineJobFinished
 691   │ === RUN   TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set
 692   │ I1228 23:01:11.596824   28776 job_controller.go:199] Starting job controller
 693   │ I1228 23:01:11.596842   28776 shared_informer.go:314] Waiting for caches to sync for job
 694   │ I1228 23:01:11.596854   28776 shared_informer.go:321] Caches are synced for job
 695   │ I1228 23:01:11.597018   28776 job_controller.go:516] enqueueing job default/job1
 696   │ I1228 23:01:11.597305   28776 job_controller.go:516] enqueueing job default/job1
 697   │ I1228 23:01:11.597547   28776 job_controller.go:516] enqueueing job default/job1
 698   │     job_controller_test.go:1908: Job failed to enforce activeDeadlineSeconds configuration. Expected condition with Reason 'DeadlineExceeded' was not found in {[] 2022-12-28 23:01:13 +0530 IST <nil> 0 0 0  nil 0xc0003d818c}
 699   │ === RUN   TestSyncPastDeadlineJobFinished/New_job_created_with_start_time_being_set
 700   │ I1228 23:01:14.597776   28776 job_controller.go:516] enqueueing job default/job2
 701   │ I1228 23:01:14.597824   28776 job_controller.go:516] enqueueing job default/job1
 702   │ I1228 23:01:14.597854   28776 job_controller.go:516] enqueueing job default/job2
 703   │ I1228 23:01:14.597868   28776 event.go:294] "Event occurred" object="default/job1" fieldPath="" kind="Job" apiVersion="" type="Warning" reason="DeadlineExceeded" message="Job was active longer than specified deadline"
 704   │ I1228 23:01:14.597878   28776 event.go:294] "Event occurred" object="default/job2" fieldPath="" kind="Job" apiVersion="" type="Warning" reason="DeadlineExceeded" message="Job was active longer than specified deadline"
 705   │ --- FAIL: TestSyncPastDeadlineJobFinished (3.30s)
 706   │     --- FAIL: TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set (3.00s)
 707   │     --- PASS: TestSyncPastDeadlineJobFinished/New_job_created_with_start_time_being_set (0.20s)
 708   │ I1228 23:01:14.797990   28776 job_controller.go:213] Shutting down job controller
```

harshanarayana replied:
Other option is to just advance the clock by `3s` and set the `ActiveDeadlineSeconds` to be `2s`. But that is not the right way to test I think

harshanarayana replied:
```bash
=== RUN   TestSyncPastDeadlineJobFinished
=== RUN   TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set
I1228 23:17:17.430555   33419 job_controller.go:199] Starting job controller
I1228 23:17:17.430569   33419 shared_informer.go:314] Waiting for caches to sync for job
I1228 23:17:17.430580   33419 shared_informer.go:321] Caches are synced for job
I1228 23:17:17.430660   33419 job_controller.go:516] enqueueing job default/job1
I1228 23:17:17.430748   33419 job_controller.go:1344] "Timing Information" job="job1" duration="0s" allowed="2s" start="2022-12-28 23:17:17 +0530 IST" current="2022-12-28 23:17:17 +0530 IST"
I1228 23:17:17.431045   33419 job_controller.go:1344] "Timing Information" job="job1" duration="0s" allowed="2s" start="2022-12-28 23:17:20 +0530 IST" current="2022-12-28 23:17:20 +0530 IST"
I1228 23:17:17.431056   33419 job_controller.go:516] enqueueing job default/job1
I1228 23:17:17.431100   33419 job_controller.go:516] enqueueing job default/job1
I1228 23:17:17.431131   33419 job_controller.go:1344] "Timing Information" job="job1" duration="0s" allowed="2s" start="2022-12-28 23:17:20 +0530 IST" current="2022-12-28 23:17:20 +0530 IST"
I1228 23:17:17.431169   33419 job_controller.go:1344] "Timing Information" job="job1" duration="0s" allowed="2s" start="2022-12-28 23:17:20 +0530 IST" current="2022-12-28 23:17:20 +0530 IST"
    job_controller_test.go:1908: Job failed to enforce activeDeadlineSeconds configuration. Expected condition with Reason 'DeadlineExceeded' was not found in {[] 2022-12-28 23:17:20 +0530 IST <nil> 0 0 0  nil 0xc00078812c}
=== RUN   TestSyncPastDeadlineJobFinished/New_job_created_with_start_time_being_set
I1228 23:17:20.431293   33419 job_controller.go:516] enqueueing job default/job2
I1228 23:17:20.431313   33419 job_controller.go:1344] "Timing Information" job="job1" duration="3s" allowed="2s" start="2022-12-28 23:17:20 +0530 IST" current="2022-12-28 23:17:23 +0530 IST"
I1228 23:17:20.431372   33419 job_controller.go:1344] "Timing Information" job="job2" duration="4s" allowed="2s" start="2022-12-28 23:17:19 +0530 IST" current="2022-12-28 23:17:23 +0530 IST"
I1228 23:17:20.431379   33419 job_controller.go:516] enqueueing job default/job1
I1228 23:17:20.431419   33419 job_controller.go:516] enqueueing job default/job2
I1228 23:17:20.431417   33419 event.go:294] "Event occurred" object="default/job1" fieldPath="" kind="Job" apiVersion="" type="Warning" reason="DeadlineExceeded" message="Job was active longer than specified deadline"
I1228 23:17:20.431432   33419 event.go:294] "Event occurred" object="default/job2" fieldPath="" kind="Job" apiVersion="" type="Warning" reason="DeadlineExceeded" message="Job was active longer than specified deadline"
--- FAIL: TestSyncPastDeadlineJobFinished (3.30s)
    --- FAIL: TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set (3.00s)
    --- PASS: TestSyncPastDeadlineJobFinished/New_job_created_with_start_time_being_set (0.20s)
```

harshanarayana replied:
I think the overall interval also needs to be much higher for polling. 3 second max poll is not good enough for some cases either it seems like. So I should bump that up to 5s may be ? @mimowo

harshanarayana replied:
```bash
=== RUN   TestSyncPastDeadlineJobFinished
=== RUN   TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set
I1228 23:23:03.546069   35573 job_controller.go:199] Starting job controller
I1228 23:23:03.546095   35573 shared_informer.go:314] Waiting for caches to sync for job
I1228 23:23:03.546113   35573 shared_informer.go:321] Caches are synced for job
I1228 23:23:03.546228   35573 job_controller.go:516] enqueueing job default/job1
I1228 23:23:03.546568   35573 job_controller.go:516] enqueueing job default/job1
I1228 23:23:03.546705   35573 job_controller.go:516] enqueueing job default/job1
    job_controller_test.go:1908: Job failed to enforce activeDeadlineSeconds configuration. Expected condition with Reason 'DeadlineExceeded' was not found in {[] 2022-12-28 23:23:05 +0530 IST <nil> 0 0 0  nil 0xc000a144dc}
=== RUN   TestSyncPastDeadlineJobFinished/New_job_created_with_start_time_being_set
I1228 23:23:08.548244   35573 job_controller.go:516] enqueueing job default/job2
I1228 23:23:08.548359   35573 job_controller.go:516] enqueueing job default/job1
I1228 23:23:08.548421   35573 event.go:294] "Event occurred" object="default/job1" fieldPath="" kind="Job" apiVersion="" type="Warning" reason="DeadlineExceeded" message="Job was active longer than specified deadline"
I1228 23:23:08.548432   35573 job_controller.go:516] enqueueing job default/job2
I1228 23:23:08.548453   35573 event.go:294] "Event occurred" object="default/job2" fieldPath="" kind="Job" apiVersion="" type="Warning" reason="DeadlineExceeded" message="Job was active longer than specified deadline"
--- FAIL: TestSyncPastDeadlineJobFinished (5.30s)
    --- FAIL: TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set (5.00s)
    --- PASS: TestSyncPastDeadlineJobFinished/New_job_created_with_start_time_being_set (0.20s)
```

Even that is giving false values on loop test. The safest is to switch to using syncJob call in this PR itself. Otherwise, I don't think we will get rid of the flakes fully.

mimowo replied:
It should be ok to bump the time to 5s.

mimowo replied:
I see that we still need the periodic poll due to cache, but do we still need `manager.clock.Sleep(200 * time.Millisecond)` and if so - can you explain why.

mimowo commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -3694,7 +3708,9 @@ func TestWatchJobs(t *testing.T) {
 	defer close(stopCh)
 	sharedInformerFactory.Start(stopCh)
-	go manager.Run(context.TODO(), 1)
+	ctx, cancel := context.WithCancel(context.Background())
```
Revert these changes, they are have nothing to do with the test. +1 for another PR for cleanup, but coupling unnecessary changes will make this important PR take more time.

harshanarayana replied:
/Ack. Sorry, this was my bad. I had this change and got pushed into this un-intentionally. It was a stash from previous thing I was suspecting. Cleaning up

mimowo commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -3740,7 +3756,9 @@ func TestWatchPods(t *testing.T) {
 	defer close(stopCh)
 	go sharedInformerFactory.Core().V1().Pods().Informer().Run(stopCh)
-	go manager.Run(context.TODO(), 1)
+	ctx, cancel := context.WithCancel(context.Background())
```
As above.

mimowo commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -3763,7 +3781,9 @@ func TestWatchOrphanPods(t *testing.T) {
 	go podInformer.Run(stopCh)
 	cache.WaitForCacheSync(stopCh, podInformer.HasSynced)
-	go manager.Run(context.TODO(), 1)
+	ctx, cancel := context.WithCancel(context.Background())
```
As above.

mimowo reviewed (commented):
IIUC the periodic checks are no longer needed as you invoke `syncJob` directly. Also, please revert the unrelated cleanups to other unit tests. Still, I'm leaning towards splitting the commits into two PRs to make the fix sink faster.

mimowo commented:
@harshanarayana if you can the PR in the current form fixes the flakes on your local setup then it lgtm.

harshanarayana commented:
> @harshanarayana if you can the PR in the current form fixes the flakes on your local setup then it lgtm.

Changing only the `ActiveDeadlinSecond` and sleep interval did not fix the test flakes fully. I still see the test flakes and they seem to be pointing to certain anomalies on how the fake tests client and controller interaction is being done. 

```bash
=== RUN   TestSyncPastDeadlineJobFinished
=== RUN   TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set
I1228 23:57:34.428505   43713 job_controller.go:698] "Job" status={Conditions:[] StartTime:<nil> CompletionTime:<nil> Active:0 Succeeded:0 Failed:0 CompletedIndexes: UncountedTerminatedPods:nil Ready:<nil>}
I1228 23:57:34.428534   43713 job_controller.go:755] Setting Start time
I1228 23:57:34.428553   43713 job_controller.go:1347] "Timing Information" job="job1" current="2022-12-28 23:57:34 +0530 IST" start="2022-12-28 23:57:34 +0530 IST" duration="0s" allowed="2s"
I1228 23:57:34.428633   43713 job_controller_test.go:1897] "Clock Advanced" current="2022-12-28 23:57:37 +0530 IST" start="2022-12-28 23:57:34 +0530 IST"
I1228 23:57:34.428689   43713 job_controller.go:698] "Job" status={Conditions:[] StartTime:<nil> CompletionTime:<nil> Active:0 Succeeded:0 Failed:0 CompletedIndexes: UncountedTerminatedPods:nil Ready:<nil>}
I1228 23:57:34.428711   43713 job_controller.go:755] Setting Start time
I1228 23:57:34.428724   43713 job_controller.go:1347] "Timing Information" job="job1" current="2022-12-28 23:57:37 +0530 IST" start="2022-12-28 23:57:37 +0530 IST" duration="0s" allowed="2s"
I1228 23:57:34.428797   43713 job_controller.go:698] "Job" status={Conditions:[] StartTime:<nil> CompletionTime:<nil> Active:0 Succeeded:0 Failed:0 CompletedIndexes: UncountedTerminatedPods:nil Ready:<nil>}
I1228 23:57:34.428814   43713 job_controller.go:755] Setting Start time
I1228 23:57:34.428826   43713 job_controller.go:1347] "Timing Information" job="job1" current="2022-12-28 23:57:37 +0530 IST" start="2022-12-28 23:57:37 +0530 IST" duration="0s" allowed="2s"
I1228 23:57:34.428863   43713 job_controller.go:516] enqueueing job default/job1
I1228 23:57:34.428876   43713 job_controller.go:516] enqueueing job default/job1
I1228 23:57:34.428883   43713 job_controller.go:516] enqueueing job default/job1
```

If you see the above, the start time attribute that is set on the resource under the status block goes missing in some cases that is causing the start and current time to be set to the current time and that is leading to test flakes. The Changes in the Current state that I have published to the PR seem to take the flakes out entirely on my local setup. 

```bash
for attempt in $(seq 1 60); do
    go test -run ^TestSyncPastDeadlineJobFinished$ k8s.io/kubernetes/pkg/controller/job -v -count 100 2>&1 > /tmp/${attempt}-test.log
done
```
Running test like above to confirm if I see flakes.

mimowo commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -1876,15 +1878,12 @@ func TestSyncPastDeadlineJobFinished(t *testing.T) {
 			if err != nil {
+				t.Errorf("failed to perform sync job operation: %v", err)
+			}
+			j, err := clientset.BatchV1().Jobs(metav1.NamespaceDefault).Get(ctx, job.GetName(), metav1.GetOptions{})
```
I'm wondering if you need to wait for `j.Status.StartTime` in a loop as it could be cached in the client instance.

harshanarayana replied:
One thing I noticed is, this part of the code never flaked while testing locally.

alculquicondor replied:
clientset doesn't use the cache, so it would be safe without a loop.

harshanarayana commented:
/test pull-kubernetes-unit

harshanarayana commented:
/test pull-kubernetes-unit

alculquicondor commented:
Will review after @mimowo adds an LGTM

mimowo commented:
/lgtm
Tested on my local env. with over 200 repeats.

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -1861,13 +1859,17 @@ func TestSyncPastDeadlineJobFinished(t *testing.T) {
+			}
+			job.Spec.ActiveDeadlineSeconds = pointer.Int64(2)
 			if tc.setStartTime {
 				start := metav1.NewTime(fakeClock.Now().Add(-time.Second))
```
I think we should just be using `Now` here.

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -1861,13 +1859,17 @@ func TestSyncPastDeadlineJobFinished(t *testing.T) {
+			if err != nil {
+				t.Errorf("failed to generate job key for performing sync job: %v", err)
+			}
+			job.Spec.ActiveDeadlineSeconds = pointer.Int64(2)
```
just keep it at 1.

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -1893,16 +1892,18 @@ func TestSyncPastDeadlineJobFinished(t *testing.T) {
-			manager.clock.Sleep(time.Second)
-			err = wait.Poll(200*time.Millisecond, 3*time.Second, func() (done bool, err error) {
-				j, err = clientset.BatchV1().Jobs(metav1.NamespaceDefault).Get(ctx, job.GetName(), metav1.GetOptions{})
+			manager.clock.Sleep(3 * time.Second)
```
```suggestion
			manager.clock.Sleep(time.Second + time.NanoSecond)
```

alculquicondor commented:
/close
This was solved in #114876

alculquicondor commented:
Thanks for the helpful debugging. With all the information gathered in this PR, I managed to find a smaller fix.
